### ISSUE TITLE
[FEATURE REQ] Better support for batch processing in Event Hubs SDK

### ISSUE THREAD

**Issue by GeeWee** [Azure/azure-sdk-for-net#24075]
**Library or service name.**
[Azure.Messaging.EventHubs]
 
**Is your feature request related to a problem? Please describe.**
(This issue is prompted by the [following twitter thread](https://twitter.com/Jesse_Squire/status/1438594052618850314))

Batch-processing when receiving events in the new EventHubs SDK is much, much more difficult than in the previous SDK. When switching over I had to re-implement a lot of the functionality you used to get out of the box.

# My batching workflow.
Just to sum up what sort of workflow I'm working with in regards to batching.
- I would like to process messages in batches of _n_.
- If _n_ messages haven't been received within 1sec, I'd like to batch process the events that are already lying around. (We have some data sources that are intermittent and we can't just leave messeages around because new ones aren't coming in)

I've followed the example [here](https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/eventhub/Azure.Messaging.EventHubs.Processor/samples/Sample04_ProcessingEvents.md#batch-processing) in regards to batch processing lead me to the following issues.

## Problem: Managing Partition State
We have to manually manage state per-partition. You used to be able to have the SDK do that for you for batching cases.
In the example listed above it seems pretty easy - you just add a ConcurrentDictionary with the partition id and whatever you want to store right? 
However let's take an example where it's not so simple.
- Processor 1 acquires partition 1 and batches up 10 messages. This is below the processing limit.
- Processor 1 loses partition 1. The messages are still batched.
- At some later point in time, Processor 1 reacquires partition 1. The old messages that are left will then be reprocessed.

To avoid these sort of circumstances you'll need to listen to all of the `PartitionClosing` events and ensure that you synchronize your state with it.

## Problem: Heartbeat messages do not allow checkpointing.
Next up - the case where I want to process my messages after _n_ seconds without any activity. Luckily there's the heartbeat message for this.

Unfortunately the heartbeat message doesn't allow us to do e.g. checkpointing or read `lastEnqueuedTime`, so I have to build up a structure that forces me to retrieve that from the last message I've enqueued.

```csharp
private async Task ProcessHeartbeatEvent(IProcessEventArgs args, string partitionId, ICheckpointer checkpointer)
{
var data = _partitionedMessageBatcher.Drain(partitionId);
// If there is already no messages on the partition, this means we've gotten multiple
// heartbeat messages in a row, and there's no need to send a list of empty
// messages any further.
if (data.Count == 0)
{
    return;
}

// The updateCheckpoint we get from Azure Event Hub are coupled to the event.
// When batching we always provide the updateCheckpoint from the latest event, except
// for when we receive heartbeat messages, in which case we use the updateCheckPoint from the last "real" event
var lastRealEvent = data.Last();
Func<CancellationToken, Task> lastUpdateCheckpointAsync = lastRealEvent.UpdateCheckpointAsync;
var lastPartitionContext =
    lastRealEvent
        .Partition; // Use partitionContext from last event as well, as the heartbeat message doesn't have the correct properties such as LastEnqueuedTime


var eventData = data
    .Select(args =>
    {
        Debug.Assert(args.Data != null,
            "args.Data != null"); // Args with null data shouldn't make it into the batcher
        return args.Data!;
    })
    .ToList();

Log.Debug("Flushing {messageCount} messages due to heartbeat message", data.Count);
var receivedEventDataBatch =
    new ReceivedEventDataBatch(lastPartitionContext, eventData, lastUpdateCheckpointAsync);
await _processEvent(receivedEventDataBatch, checkpointer);
}
```

Perhaps this is because my codebase is shaped by the old SDK where checkpointing was done on a batch basis rather than as a function provided by each event received.

# Summing up
- Batching is harder than it needs to be. I imagine this use-case is common and I would prefer if something was provided that helped you do batching, similar to the old-style SDK.

- However I think in lieu of that, something that will help you manage partition state would be nice. I'm not quite sure how that would work however.

- I would also like if you were able to call `UpdateCheckpointAsync` on a heartbeat message, and that would then checkpoint all the previous messages.

- If nothing else, a more involved example in the documentation would be nice.

**Comment by jsquire** [comment]
Hi @GeeWee.   Thank you for your feedback and we regret that you're experiencing difficulties.  The efforts that you put into providing this level of detail on your scenarios is greatly appreciated.

From your scenario, there are definitely gaps in the functionality provided by the `EventProcessorClient` and at a high level, it your approach works around them in the way that I would recommend.  That said, I believe that it may not be the best type for your needs.

## `EventProcessorClient` versus `EventProcessor<T>`

For a bit of context, the role that the `EventProcessorClient` plays is a bit different than the `EventProcessorHost` from the previous SDK.  `EventProcessorClient` was designed to provide a familiar API to that of the `EventHubConsumerClient` and offer an intuitive "step-up" experience for developers exploring Event Hubs as they advance to production scenarios.  For a large portion of our library users, that covers their needs well.   There's definitely a point, however, where an application requires more control to handle higher throughput or more specialized needs - that's where the `EventProcessor<T>` is intended to help.
 
The `EventProcessor<T>` fills a similar role as the `EventProcessorClient`, with cooperative load balancing and resiliency as its core features but also offers native batch processing, extension points for influencing load balancing, customizing of storage providers, and a lighter, less opinionated API.  The caveat is that this comes in the form of an abstract base, which needs to be extended and the core “handler” activities implemented via override.  More on the design and philosophy behind this can be found [here](https://github.com/Azure/azure-sdk-for-net/blob/master/sdk/eventhub/Azure.Messaging.EventHubs/design/proposal-event-processor%7BT%7D.md).

## `EventProcessor<T>` Considerations

The design more closely matches that of the legacy EPH.  You're working with batches natively able to define a class similar to the legacy `IEventProcessor` as `T` which has an instance per partition and is a good place for storing partition-specific state and methods.

The major downside to using the `EventProcessor<T>` is that there is no storage integration provided.  This requires that an application implement four key storage operations:

  - Get a Checkpoint
  - Upsert a Checkpoint
  - List Ownership
  - Update Ownership

While these are fairly straight-forward, it is tedious.  This can also be a burden if you're trying to integrate into a heterogeneous environment where other `EventProcessorClients` may be running and you'd like your data to be compatible with those.

If you're interested in exploring further, [here](https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/eventhub/Azure.Messaging.EventHubs/samples/Sample08_CustomEventProcessor.md) is a sample for implementing a custom processor.  There is also an article that offers deeper discussion [here](https://devblogs.microsoft.com/azure-sdk/custom-event-processor/).   The storage implementation for the `EventProcessorClient` is open source and can be found [here](https://github.com/Azure/azure-sdk-for-net/tree/main/sdk/eventhub/Azure.Messaging.EventHubs.Shared/src/BlobCheckpointStore).

## Doing Better

There are parts of this experience that I, personally, do not like and think that we could do better.   I'm going to offer some additional thoughts organized into individual comments below.  My intent is to allow interested parties to up-vote them individually and make it a bit easier to read.

**Comment by jsquire** [comment]
# Publish Storage Manager Implementations

The main drawback that I currently see with `EventProcessor<T>` is the lack of any out-of-the-box storage integration options.  In previous SDKs, it was possible to plug-in a new storage implementation and avoid taking on that concern.  

There is a bit of friction in the current design for this being a complete drop-in, but I believe that offering an abstract base that defines a common structure and then exposing the concrete implementation alongside the `EventProcessorClient` in each package would ease the burden that custom processor implementations face and potentially open up an ecosystem for community-developed storage packages.

## Example `StorageManager` Structure

```csharp
public abstract class StorageManager
{
    public abstract Task<IEnumerable<EventProcessorPartitionOwnership>> ListOwnershipAsync(
        string fullyQualifiedNamespace,
        string eventHubName,
        string consumerGroup,
        CancellationToken cancellationToken);

    public abstract Task<IEnumerable<EventProcessorPartitionOwnership>> ClaimOwnershipAsync(
        IEnumerable<EventProcessorPartitionOwnership> partitionOwnership,
        CancellationToken cancellationToken);

    public abstract Task<EventProcessorCheckpoint> GetCheckpointAsync(
        string fullyQualifiedNamespace,
        string eventHubName,
        string consumerGroup,
        string partitionId,
        CancellationToken cancellationToken);

    public abstract Task UpdateCheckpointAsync(
        string fullyQualifiedNamespace,
        string eventHubName,
        string consumerGroup,
        string partitionId,
        long? offset,
        long? sequenceNumber,
        CancellationToken cancellationToken);
} 

// These exist today and are shown for context because the storage manager references them.

public class EventProcessorPartitionOwnership
{
    public string ConsumerGroup { get; set; }
    public string EventHubName { get; set; }
    public string FullyQualifiedNamespace  { get; set; }
    public DateTimeOffset LastModifiedTime  { get; set; }
    public string OwnerIdentifier  { get; set; }
    public string PartitionId { get; set; }
    public string Version  { get; set; }
}

public class EventProcessorCheckpoint
{
    public string FullyQualifiedNamespace { get; set; }
    public string EventHubName { get; set; }
    public string ConsumerGroup { get; set; }
    public string PartitionId { get; set; }
    public EventPosition StartingPosition { get; set; }
}
```

This is reasonably close to the form that is used internally today, and would give custom processors an interface to wire up in the storage method stubs that need to be overridden, as well as allow for checkpoint creation with a minimal set of data rather than the need to hold onto an `EventData` instance.

**Comment by jsquire** [comment]
# Graft Batch Processing onto the `EventProcessorClient`

This would require extending the existing `EventProcessorClient` and supporting types with some new members to allow for a choice between registering a handler for single dispatch (like today) or batch dispatch.   The design challenge is to do be able to support this without introducing any breaking changes to the existing API.

The main drawback that I see is that it adds complexity to the current API and makes the "step-up" experience from the consumer client a bit less clear.   I am also not convinced that it goes far enough to address some of the concerns about flexibility for checkpointing, as there would still be an `EventData` instance needed.

## Example:  Extended `EventProcessorClient`

```csharp
public class EventProcessorClient : EventProcessor<EventProcessorPartition>
{
    // Now there are two handlers; only one can be registered at a time.

    public event Func<ProcessEventArgs, Task> ProcessEventAsync;
    public event Func<ProcessEventBatchArgs, Task> ProcessEventBatchAsync;

    // Existing event handlers

    public event Func<PartitionInitializingEventArgs, Task> PartitionInitializingAsync;
    public event Func<PartitionClosingEventArgs, Task> PartitionClosingAsync;
    public event Func<ProcessErrorEventArgs, Task> ProcessErrorAsync;

    // Existing members for the processing context

    public string FullyQualifiedNamespace { get; }
    public string EventHubName { get; }
    public string ConsumerGroup { get; }
    public string Identifier { get; }
    public bool IsRunning { get; }

    // ... SNIP (constructors, start and stop)
}

// New args for the batch handler; follows the existing pattern for 
// single dispatch to keep things as similar as possible.

public struct ProcessEventBatchArgs
{
    public bool HasEvents { get; }
    public PartitionContext Partition { get; }
    public IReadOnlyList<EventData> EventBatch { get; }
    public CancellationToken CancellationToken { get; }

    public Task UpdateCheckpointAsync(
        EventData checkpointEvent,
        CancellationToken cancellationToken = default);
}

// Existing options with a new/hide to rename an existing 
// member to better fit the batch context.  
// 
// The old member will continue to exist, but be hidden from 
// Intellisense and delegate to the new member.

public class EventProcessorClientOptions
{
    // New member to replace the cache event count; this will be
    // documented to highlight that it is either used as the
    // maximum batch size for bach processing or the amount of
    // events read and cached for single-dispatch.

    public int MaximumBatchEventCount { get; set; }

    // Hide the existing member, as the name may be confusing when
    // the batch event handler is used.

    [EditorBrowsable(EditorBrowsableState.Never)]
    public int CacheEventCount
    {
        get => MaximumBatchEventCount;
        set => MaximumBatchEventCount = value;
    }

    // ... SNIP (existing options members)
}
```

Personally, I'm not a fan of this.  I believe that it clutters the API and muddies the "step-up" experience, but only solves part of the problem.  If we were to remove friction from extending `EventProcessor<T>`,  I think that would better address the more more advanced cases and still be approachable for the simple "I just want batch processing" scenario.

**Comment by GeeWee** [comment]
Hey @jsquire - thanks for the extremely detailed feedback and considerations.

I have a few thoughts:

## Documentation
I honestly think a lot of this could probably be solved by documentation that leads the user to make the right choices:

- I generally agree that the dichotomy of `EventProcessorClient` vs `EventProcessor` makes sense. Perhaps this is a documentation issue, but I recall reading about the `EventProcessor` and thinking "Oof, plug in my own storage implementation? That's not what I'm looking for" - however it probably was exactly what I was looking for in regards to batching.
- I think it can be made more explicit that you're probably looking for the `EventProcessor` if you want advanced use-cases and batching - even though you might not be interested in plugging your own storage implementation.
- Particularly the sample [here](https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/eventhub/Azure.Messaging.EventHubs.Processor/samples/Sample04_ProcessingEvents.md#batch-processing) lead me to believe that I was meant to use the `EventProcessorClient` for batching scenarios.
- All of the articles and link you post to design documents and articles make a lot of sense - I wasn't able to find them on my own. Perhaps this is ineptitude or perhaps they're not highlighted enough.


## Technical
In regards to the stuff you've posted I 100% agree that the right call is to make the `EventProcessor` easier to extend from, rather than grafting event processing onto the `EventProcessorClient`

I think there are a couple of options you can take. Either you can document how the Azure blob storage StorageManager works so people can copy that implementation, or you can define a `StorageManager` like you suggested - and then perhaps allow people to use the "standard" one used in the `EventProcessorClient`

**Comment by amshalev** [comment]
Hi @jsquire, is there any update\ETA regarding this request?
I'm working on migrating our code (Kusto's EventHub data connection) from EventProcessorHost to EventProcessorClient.
I have the exact same problems which @GeeWee described, our code and design based on batch processing, which is missing in the new SDK, so currently it's blocking the migration.

**Comment by jsquire** [comment]
Hi @amshalev.   No updates at this time.   This will take some time to move forward, as it requires multiple rounds of internal discussion and design proposals before we'll have an understanding of the outcome.  At that point, any changes would need to be prioritized and scheduled.  The bottom line is that, though this is something that I'd really like to do, it'll be a longer-term effort.

I do want to mention that bath processing is not missing, it comes in a different form that requires using `EventProcessor<T>` rather than the `EventProcessorClient`.  The guidance [above](https://github.com/Azure/azure-sdk-for-net/issues/24075#issuecomment-921795041) would be my recommendation for Kusto as well.

**Comment by amshalev** [comment]
Thanks for the quick reply!
I see... the problem with using the `EventProcessor<T>` is that it requires to rewrite\copy a lot of code which already exists in `EventProcessorClient`, however, after looking on the code it seems that using the `BlobsCheckpointStore` will make it much easier. 
But I was disappointed to find that `BlobsCheckpointStore` class marked as `internal`, could you consider change it to be public so it will be possible to consume it directly?

**Comment by jsquire** [comment]
Unfortunately, we can't just flip that type to public.  Anything that we make part of the public API has to go through an API approval process and review, and comes with a strong backwards compatibility guarantee.    The approach that we'd need to take for that path is the first option that I wrote up above - and what would by my preferred path.

The existing `BlobCheckpointStore` implementation is open source;  you're free to pull the code into your project as a copy or submodule - though the caveat with the latter is that we'll continue to view that as an internal type and may make breaking changes.

**Comment by amshalev** [comment]
I'm aware that I can pull this code to my project, but I don't think that's a right and maintainable solution.
I guess that we will stay with `EventProcessorHost` for now and will wait for progress of one of the options you've suggested.
Thanks

**Comment by SanderSade** [comment]
The design of not having the batch processing out-of-the-box like the previous version seems to be rather odd.

The event handler (ProcessEventAsync) is counter-intuitive to use, especially in light of high-performance scenarios for which Event Hubs are created. Batch processing is central in writing events to the hub, so it would be only logical to have reading the events in a batch as the default option as well.

I am forced to stay to EventProcessorHost, too - either I duplicate a lot of code from EventProcessorClient and related files - or cannot use the batch processing. And when there is a new Azure.Messaging.EventHubs.Processor release, then I would have to backport all the changes into my code, hoping that I haven't done any breaking changes to that code. This is not really an optimal way to handle event processing.

**Comment by jsquire** [comment]
Hi folks,

Apologies for the lack of updates.  Thanks in no small part to the discussion on this issue, we were able to prove the need for a better story around extending the processor and batch support.   Starting with our next release, v5.7.0-beta.5, we've made the following improvements:

- The `Azure.Messaging.EventHubs` package now defines a `CheckpointStore` type to normalize processor storage operations.

- The `Azure.Messaging.EventHubs` package includes a `PluggableCheckpointStoreEventProcessor<T>` that can be extended with your processing logic without the need to implement storage operations.

- The Blob Storage implementation used by the `EventProcessorClient` is now public in the `Azure.Messaging.EventHubs.Processor` as `BlobCheckpointStore`, and can be used when extending processor types.

- All event processor types now expose a protected `UpdateCheckpoint` member that can types extending them can call.  This new method does not require an `EventData` instance to create the checkpoint, only an offset.

More details can be found in [this sample](https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/eventhub/Azure.Messaging.EventHubs/samples/Sample08_CustomEventProcessor.md#building-the-custom-event-processor).
